### ISSUE TITLE
CSS link helper tag with fallback class not working with subclasses

### ISSUE THREAD

**Issue by johnwc** [aspnet/Mvc#6264]
The **bootstrap-dialog** stylesheet does not have any classes that are single non subclasses. The below taghelper is not working.

```html
<link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/bootstrap3-dialog/1.35.4/css/bootstrap-dialog.min.css"
      asp-fallback-href="~/lib/bootstrap3-dialog/dist/css/bootstrap-dialog.min.css"
      asp-fallback-test-class="bootstrap-dialog modal-header"
      asp-fallback-test-property="border-top-left-radius"
      asp-fallback-test-value="4px" />
```

**Comment by Eilon** [comment]
Hi @johnwc , I'm not sure I understand the issue. Can you explain what you mean by "single non subclasses" and what problem that's causing?

**Comment by johnwc** [comment]
If you look at the stylesheet you will see that there is no single class, they are all subclasses. There is no class `.bootstrap-dialog {...}` but only `.bootstrap-dialog .modal-header {...}` How do you tell the script to test for classes like that?

**Comment by Eilon** [comment]
Ah got it, thanks!

@DamianEdwards - any thoughts on how to check this from the script?

**Comment by johnwc** [comment]
[dataTables.bootstrap.css](https://cdnjs.cloudflare.com/ajax/libs/datatables/1.10.15/css/dataTables.bootstrap.css) stylesheet is another good example. All of it's styles are element specific, either table or div.

**Comment by Eilon** [comment]
I wonder if there even exists a good solution to this? I'm not much of an expert when it comes to the problem of how to detect whether some CSS failed to load...

@javiercn / @sebastienros - perhaps you have some ideas on how you could detect whether CSS failed to load, when it only has more complex classes?

**Comment by johnwc** [comment]
It'd be ideal to create a standardized "test" class that stylesheet developers start using. So that every stylesheet have a unique named class just for testing if it's loaded. All it would have/need is a `display: none`.

Example:
```css
.bootstrap-dialog-loaded {
   display: none;
}
```

**Comment by DamianEdwards** [comment]
There's no good way I know of to handle testing whether it loaded and applied if it doesn't have a suitable class that can be used as a "marker".

A quick search shows that `document.styleSheets` may be a more direct approach we could take: http://stackoverflow.com/questions/4724606/how-to-use-javascript-to-check-and-load-css-if-not-loaded#answer-25615777

This would involve us changing from testing whether a known CSS class's rules have been applied to the generated `<meta />` element, to directly inspecting via the DOM whether the stylesheet has loaded or not.

**Comment by mkArtakMSFT** [comment]
@DamianEdwards, do you think we need to take any action here? If not, just close please.

**Comment by johnwc** [comment]
Yes, action needs to be taken to fix the issue to allow testing for subclasses.

**Comment by mkArtakMSFT** [comment]
@NTaylorMullen, @SteveSandersonMS can you guys please work together to come up with a good solution here? Thanks!

**Comment by NTaylorMullen** [comment]
Looking at `document.styleSheets` seems to be a great fix for this and would make the css link TagHelper far more useful. @SteveSandersonMS may have more context on if there's a better way to fix this issue, or if there's hidden pitfalls in `document.styleSheets` that we're unaware of.

**Comment by mkArtakMSFT** [comment]
@NTaylorMullen, @SteveSandersonMS, wouldn't this be a better option: https://github.com/aspnet/Mvc/issues/6619

**Comment by johnwc** [comment]
No, that's not a option. That issue actually sounds like the developer doesn't understand how the fall back test works at all. He just needs to follow what was suggested and use a unique style class for that specific stylesheet.

**Comment by mkArtakMSFT** [comment]
This is not a priority at the moment nor we have received much community interest on this issue.
 Hence I'd be happy to decide the approach we'd like to take here and leave it aside for the community to contribute to, if they will need it. @johnwc, will you be able interested in submitting a PR for this?
@SteveSandersonMS, would be great if you could provide your feedback regarding the proposed solution by @NTaylorMullen.
> Looking at document.styleSheets seems to be a great fix for this and would make the css link TagHelper far more useful. @SteveSandersonMS may have more context on if there's a better way to fix this issue, or if there's hidden pitfalls in document.styleSheets that we're unaware of.

**Comment by SteveSandersonMS** [comment]
Yes, I agree that inspecting `document.styleSheets` is most likely the most direct solution here.

If we really wanted to keep the existing "feature detection"-style approach, we could change the API to let developers provide the ID of an element that matches one of the selectors used in the stylesheet, then we check whether the expected style property has been applied to it. But checking `document.styleSheets` would be simpler.

**Comment by johnwc** [comment]
Yes, I'd be more than happy to do a PR for this and put some volunteer time in for the project.

What approach are we looking to update it to, the `document.styleSheets` version?

**Comment by mkArtakMSFT** [comment]
Hi @johnwc.
Sorry we've missed your last comment.
Are you still interested in submitting a PR for this? The approach would be the `document.styleSheets` way @SteveSandersonMS has detailed in his comment.
